include/simulation_time/simulation_clock.hpp

```diff
@@ -0,0 +1,205 @@
+    using calendar_time_point = calendar_clock::time_point;
+
+    // simulation-calendar conversion functions //
+
+    //! Convert a calendar time point to a simulation time point.
+    //! \param calendar_time Calendar time point.
+    //! \returns The corresponding simulation time point.
+    //! \pre simulation_clock::set_epoch must have been called prior to calling this.
+    static simulation_time_point from_calendar(const calendar_time_point& calendar_time);
+
+    //! Convert a simulation time point to a calendar time point.
+    //! \param sim_time Simulation time point.
+    //! \returns The corresponding calendar time point.
+    //! \pre simulation_clock::set_epoch must have been called prior to calling this.
+    static calendar_time_point to_calendar(const simulation_time_point& sim_time);
+
+    //! Convert a std::time_t value to a simulation time point.
+    //! \see simulation_clock::from_calendar
+    //! \param time Unix epoch-based time value
+    //! \returns The corresponding simulation time point.
+    //! \pre simulation_clock::set_epoch must have been called prior to calling this.
+    static simulation_time_point from_time_t(std::time_t t);
+
+    //! Convert a simulation time point to a std::time_t value.
+    //! \see simulation_clock::to_calendar
+    //! \param sim_time Simulation time point.
+    //! \returns The corresponding std::time_t value.
+    //! \pre simulation_clock::set_epoch must have been called prior to calling this.
+    static std::time_t to_time_t(const simulation_time_point& sim_time);
+
+    // simulation management functions //
+
+    //! Unitialized epoch value, aka the epoch of typename simulation_clock::calendar_clock.
+    static constexpr auto uninitialized_epoch = calendar_time_point{};
+
+    //! Get the simulation epoch.
+    //! \see simulation_clock::calendar_time_point
+    //! \see simulation_clock::set_epoch
+    //! \returns
+    //! Calendar time point representing the simulation start date/time.
+    //! If the epoch has not been set, then simulation_clock::uninitialized_epoch
+    //! is returned.
+    static calendar_time_point epoch() noexcept;
+
+    //! Set the simulation epoch.
+    //! \note This function must only be called once.
+    //! \param calendar_time Calendar time set as the epoch.
+    //! \throws std::logic_error Thrown when this function is called more than once.
+    static void set_epoch(calendar_time_point calendar_time);
+
+    //! Check if the simulation epoch has been set.
+    //! \returns true when the epoch is set, and false otherwise.
+    static bool has_epoch() noexcept;
+
+    //! Forward the simulation by N ticks.
+    //! \param steps Number of steps to tick.
+    //! \note Ticking does *not* require that simulation_clock::set_epoch has been called.
+    //! \see simulation_clock::period
+    //! \pre `steps` must be non-negative.
+    //! \throws std::domain_error Throws when the precondition is violated.
+    static void tick(rep steps = 1);
```

Comment:
Just a thought, do we need thread-safety guarantees?

> I don't think so...

Reply:
No, I don't think so. Any multi-threading we eventually, hypothetically do should be contained within a single simulation time-step, with suitable synchronization around ticks.

Reply:
Cool, that's in line with what I was thinking.